src/torrent_info.cpp

```diff
@@ -579,24 +580,30 @@ namespace {
 				return false;
 			}
 
-			std::string path = root_dir;
-			sanitize_append_path_element(path, e.first);
+			string_view filename = { e.first.data() + info_ptr_diff, e.first.size() };
+			while (!filename.empty() && filename.front() == TORRENT_SEPARATOR)
+				filename.remove_prefix(1);
+
+			bool const leaf_node = e.second.dict_size() == 1 && e.second.dict_at(0).first.empty();
+			bool const single_file = leaf_node && top_level && tree.dict_size() == 1;
 
-			if (e.second.dict_size() == 1 && e.second.dict_at(0).first.empty())
+			std::string path = single_file ? std::string() : root_dir;
```

Comment:
I turned around the logic a bit here. Instead of setting `path` unconditionally, and then clear it if we're a single file torrent, I only set it if we're not.